### DISCUSSION TITLE
Help with view collection

### DISCUSSION THREAD

**Discussion by dimitriskourg** [pocketbase/pocketbase#2373] [Q&A] [answered]
Hi,

I have two base collections. The first one is called rooms and contains the id, and the name of a room. The second one is called reservations and contains the id, a date field called fromDate, a date field called toDate and a relation field called rooms that contains all rooms of a reservation.

I want to create a view collections that I would pass a fromDate and a toDate and will return all the available rooms (ids of the rooms) at those dates based on the reservations. Do you have any idea how I could implement that using a view collection? 

Thank you

**Comment by ganigeorgiev** [comment, marked as answer]
View collections can't be parameterized (aka. you can't have dynamic `fromDate` and `toDate` as part of the view query statement).

In the near future (probably v0.17.0 release) there will be support for indirect filter and sort and you'll be able to filter rooms based on the indirect reservation relation dates. For now, one workaround could be to filter first the reservations and then send second request to fetch the non-booked rooms (_if that's your use case_). For example (_I didn't tested it_):

```js
const fromDate = "...";
const toDate   = "...";

const reservations = await pb.collection("reservations").getFullList({
    // note that dates are stored in UTC and in the format "Y-m-d H:i:s.uZ"
    filter: `fromDate >= "${fromDate}" && toDate <= "${toDate}"`,
});

// extract the ids of the booked rooms (assuming that rooms is a multiple relation field)
const bookedRooms = [];
for (let reservation of reservations) {
    for (let id of reservation.rooms) {
        if (!bookedRooms.includes(id)) {
            bookedRooms.push(id);
        }
    }
}

// fetch all other rooms (aka. rooms that don't exist in the bookedRooms array)
const freeRooms = await pb.collection("rooms").getFullList({
    filter: bookedRooms.map((id) => `id!="${id}"`).join("&&"),
})
```

**Reply by ganigeorgiev** [reply]
Note that the above may not work correctly if the date interval spans several days and a room can be "booked" and "not booked" at the same time within the date interval.

Without registering a [new route and db query](https://pocketbase.io/docs/router/#register-new-route) I'm not sure if this could be handled more easily.
Alternatively, depending on your use case and if there is a pagination with for example ~20-30 rooms, then you can also send individual reservation check requests for each room.

**Reply by ganigeorgiev** [reply]
Another workaround if the rooms are fixed is to fetch all of them upfront, send only a single reservation request for the dates and then build the available intervals for each room client-side.

But in any case, if I'm not misunderstanding your use case, I don't think this could be done easily with just a single sql query, since you'll have to somehow "merge" the available intervals within the requested period (_for example within a month a room could be booked on 1-3, 5-10 and 20-25, so the room would have an "availability range" of [4, 11-19 and 26-31] or something like that_).

**Reply by dimitriskourg** [reply]
> View collections can't be parameterized (aka. you can't have dynamic `fromDate` and `toDate` as part of the view query statement).
> 
> In the near future (probably v0.17.0 release) there will be support for indirect filter and sort and you'll be able to filter rooms based on the indirect reservation relation dates. For now, one workaround could be to filter first the reservations and then send second request to fetch the non-booked rooms (_if that's your use case_). 

Thanks, that was exactly what I was looking about! 
Also, I want to thank you for building pocketbase and I can see you've put in immense thought into making it work, and I really appreciate that.

**Comment by Nciso** [comment]
@ganigeorgiev Hi! passing a parameter to filter data is not even possible using pocketbase as a framework?

I have this query, I want to know if there is a workaround, even if I have to do a very specific thing
```
SELECT s.name AS subcategory_name, SUM(l.amount) AS total
FROM Ledger l
JOIN Subcategories s ON l.subcategory_id = s.subcategory_id
JOIN Categories c ON s.category_id = c.category_id
WHERE c.name = 'SomeCat' AND l.date >= '<QUERY_PARAM_HERE>' AND l.date <= '<QUERY_PARAM_HERE>'
GROUP BY s.name;
```

**Reply by Nciso** [reply]
AH I see, I can add a custom router and to it there 
https://pocketbase.io/docs/router/
